File: edison-togglz/src/test/java/de/otto/edison/togglz/controller/FeatureTogglesRepresentationTest.java
package de.otto.edison.togglz.controller;

import de.otto.edison.testsupport.togglz.TestFeatureManager;
import de.otto.edison.togglz.EmptyFeatures;
import de.otto.edison.togglz.TestFeatures;
import org.junit.jupiter.api.Test;
import org.togglz.core.manager.FeatureManager;
import org.togglz.core.repository.FeatureState;
import java.util.List;
import java.util.Map;

import static de.otto.edison.togglz.controller.FeatureTogglesRepresentation.togglzRepresentation;
import static org.hamcrest.MatcherAssert.assertThat;
import static org.hamcrest.Matchers.is;
import static org.hamcrest.Matchers.notNullValue;

class FeatureTogglesRepresentationTest {

    private FeatureTogglesRepresentation testee;

    @Test
    void testGetFeatureRepresentation() {
        FeatureManager featureManager = new TestFeatureManager(TestFeatures.class);
        testee = togglzRepresentation(featureManager);

        final Map<String, FeatureToggleRepresentation> features = testee.features;
        assertThat(features.get("TEST_FEATURE"), is(FeatureToggleRepresentation.newBuilder()
                .withDescription("a test feature toggle")
                .withEnabled(false)
                .build()));
    }

    @Test
    void testGetEmptyFeatureRepresentation() {
        FeatureManager featureManager = new TestFeatureManager(EmptyFeatures.class);
        testee = togglzRepresentation(featureManager);

        final Map<String, FeatureToggleRepresentation> features = testee.features;
        assertThat(features, is(notNullValue()));
        assertThat(features.isEmpty(), is(true));
    }

    @Test
    void shouldResolveFeatureGroups() {
        FeatureManager featureManager = new TestFeatureManager(TestFeatures.class);
        testee = togglzRepresentation(featureManager);

        final Map<String, FeatureToggleRepresentation> features = testee.features;
        assertThat(features.get("TEST_FEATURE_2"), is(FeatureToggleRepresentation.newBuilder()
                .withDescription("TEST_FEATURE_2")
                .withEnabled(false)
                .withGroups(List.of("TestToggleGroup"))
                .build()));
    }

    @Test
    void shouldResolveStrategy() {
        FeatureManager featureManager = new TestFeatureManager(TestFeatures.class);
        FeatureState state = new FeatureState(TestFeatures.TEST_FEATURE, true);
        state.setStrategyId("someStrategy");
        featureManager.setFeatureState(state);
        testee = togglzRepresentation(featureManager);

        final Map<String, FeatureToggleRepresentation> features = testee.features;
        assertThat(features.get("TEST_FEATURE").strategy, is("someStrategy"));
    }
}

File: edison-togglz/src/test/kotlin/de/otto/edison/togglz/controller/representation/FeatureTogglesRepresentationTest.kt
package de.otto.edison.togglz.controller.representation

import de.otto.edison.togglz.*
import de.otto.edison.togglz.controller.FeatureToggleRepresentation
import de.otto.edison.togglz.controller.FeatureTogglesRepresentation.togglzRepresentation
import io.kotlintest.shouldBe
import org.junit.jupiter.api.AfterEach
import org.junit.jupiter.api.Test

internal class FeatureTogglesRepresentationTest {
    @Test
    fun testGetFeatureRepresentationForKotlinEnumClass() {

        //given
        KFeatureManagerSupport.allEnabledFeatureConfig(KotlinTestFeatures::class)

        //when
        val togglzRepresentation = togglzRepresentation { FeatureEnum::class.java }

        //then
        val features = togglzRepresentation.features
        val fooRepresentation: FeatureToggleRepresentation? = features["FOO"]
        fooRepresentation!!.enabled shouldBe true

        val barRepresentation: FeatureToggleRepresentation? = features["BAR"]
        barRepresentation!!.enabled shouldBe true
    }

    @AfterEach
    internal fun tearDown() {
        KFeatureManagerProvider.instance = null
    }
}